### PR TITLE
remote-week1: initial commit

### PR DESCRIPTION
**Pull request by jade-f-t**
Elastic IP address : 3.218.196.210
Database host name : database-5.citefjbsojxf.us-east-1.rds.amazonaws.com
Change descriptions : 
- add README titles

### CODE REVIEW

**Review by addie-tyc — APPROVED** [review]
LGTM

**Comment by jade-f-t** [comment]
**Elastic IP address : 3.218.196.210
Database host name : database-5.citefjbsojxf.us-east-1.rds.amazonaws.com**

**<1> Assignment 1: Synchronous vs. Asynchronous Function**
- requestSync_execution_time(ms):
requestSync_execution_time(ms)

- requestAsync_execution_time(ms):
requestAsync_execution_time(ms)

- Based on the result of experiment, what's the differences of sync/async functions we can tell?

>  For sync function, it executed the code sequentially, thus , we needed to wait the first function to finish before executing the second and third function. Therefore, in the terminal, we saw that the ~5000ms appeared one by one.

> For async function, it allowed the following function to continue executing, thus, the time needed to go through the start to the end of the function was relatively short (~20,1,1 ms). Therefore, we can see that in the terminal, after the execution time of the third function showed, the program was notified that the first program finally finished. 

> Comparing the total execution time, actually they are similar ( ~5000ms ), but for sync function, the whole program will be stucked by each function.

- In which types of situations are asynchronous functions most useful?

> When the program need to do a potentially long-running task and still be responsive to other events while that task runs, async function will be useful. For example, reading files or fetching data.

- We've known the 3 ways to implement async functions, how do we choose from them and what are their pros and cons?

> For callback, it is easy to implement. Before learning the other two methods, understanding callback can make things clearer. But if we need to have multiple chain of the function, we will enter callback hell and hard to be handled.

> For promise, it solve the problem of callback hell and make the code and handle error in a more structured way. However, it is hard to manage multiple promises in a chain.

> For async/await function, it can be written in like the synchronize code, therefore it will be simpler, and it will be easier to handle the promise chain.

> Overall, the performance of promises and async/await function is better than callback, and they handle errors better, thus in most situation, they may be better. (compatibility?)

- Any other observations

> the performance of promises and async/await function is better than callback, callback needs ~20ms to execute while the other two only need ~1ms to execute. Also, the total execution time are similar ( ~5000ms ).

- third party modules : XMLHttpRequest:

> XMLHttpRequest can be used to make HTTP requests, therefore it allow me to request the URL. Another way to do it should be using fetch, but i havent used it before. After researching for the differences, i found that the performance of fetch might be better than XMLHttpRequest, thus i may try fetch afterwards. But in this case (not too much step to be finished), i guess they both have similar performance.

**<2> The first backend service**
- Change descriptions:
    - port number = 3000
    - created the first API /healthcheck

**<3> Create MySQL Database and Table**
- additional data file : backup.sql

**<4> Complete Admin APIs**
- Change descriptions:
    - port number = 3000
    - completed Sign Up API
    - completed Qeury API

**Comment by addie-tyc** [comment]
> the performance of promises and async/await function is better than callback, callback needs ~20ms to execute while the other two only need ~1ms to execute. Also, the total execution time are similar ( ~5000ms ).

結論有點怪，我看你目前的 code 好像沒有印出 20, 1, 1 的 code？
要注意單位其實是 ms，所以 20ms 幾乎是立刻馬上就印出來了，表示你發出去的 request 還沒回來就把時間印出來了，這樣的時間差可能純粹是因為你在 callback 比較晚印時間而已  

---

update: 有看到印出來的 code 了

File: package.json [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+  "main": "app.js",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
+  },
+  "keywords": [],
+  "author": "",
+  "license": "ISC",
+  "dependencies": {
+    "express": "^4.18.2",
+    "pug": "^3.0.2"
```
Comment (addie-tyc) [inline package.json:14]: 有用到 XHR 但好像沒有在 package.json 裡頭？

File: app.js [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+	res.render('healthcheck');
+});
+
+app.post('/users',(req,res)=>{
+	//get the data that the user inputted
+	const name = req.body.name;
+	const email = req.body.email;
+	const password = req.body.password;
+	//const date = req.body.date;
+	const date =new Date().toUTCString()
```
Comment (addie-tyc) [inline app.js]: response 的要求是要回傳 headers 裡的 request-date 呦 XD

File: app.js [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+				
+				const userINFO = {
+					'id': result[0].id,
+					'name': result[0].name,
+					'email': result[0].email 
+				};
+				const date = result[0].created.toUTCString()
+				const data = {
+					'user':userINFO,
+					'date':date
```
Comment (addie-tyc) [inline app.js:108]: 這裡也是，回傳的 date 是從 headers 來的呦

**Review by addie-tyc — CHANGES_REQUESTED** [review]
除了 request-date 之外沒什麼問題～

**Review by addie-tyc — APPROVED** [review]
LGTM

**Pull request merged**